File: frontend/src/types/transaction.ts
export interface FieldCoordinate {
  text: string
  x: number
  y: number
  width: number
  height: number
  bbox: [number, number, number, number]
  font: string
  font_size: number
}

export interface TransactionCoordinates {
  date?: FieldCoordinate | null
  description?: FieldCoordinate | null
  debit?: FieldCoordinate | null
  credit?: FieldCoordinate | null
  balance?: FieldCoordinate | null
}

export interface ParsedTransaction {
  transaction_id: string
  page: number
  row_index: number
  date: string | null
  description: string
  debit: string | null
  credit: string | null
  balance: string | null
  coordinates: TransactionCoordinates
  font_metadata: Record<string, unknown>
  row_bbox: [number, number, number, number]
  confidence: number
  validation_warnings: string[]
}

export interface TransactionSummary {
  total_debit: string | number
  total_credit: string | number
  opening_balance: string | number | null
  closing_balance: string | number | null
  transaction_count: number
  validation_passed: boolean
  validation_issues: string[]
}

export interface ColumnDefinition {
  name: string
  x_min: number
  x_max: number
  x_center: number
}

export interface ParseDebugInfo {
  columns: ColumnDefinition[]
  grouped_row_count: number
  raw_row_count: number
  header_row_index: number | null
  extraction_mode?: string
  layout_confidence?: number | null
  ocr_confidence?: number | null
  header_row_y?: number | null
  bank_layout_version?: string | null
}

export interface TransactionsResponse {
  statement_id: string
  bank: string
  bank_confidence: number
  transactions: ParsedTransaction[]
  summary: TransactionSummary
  cached: boolean
  warnings: string[]
  debug?: ParseDebugInfo | null
  extraction_mode?: string
  layout_confidence?: number | null
  ocr_confidence?: number | null
}

File: backend/app/schemas/transaction.py
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from app.ai_engine.models import ParseDebugInfo, TransactionSummary


class FieldCoordinateSchema(BaseModel):
    text: str
    x: float
    y: float
    width: float
    height: float
    bbox: list[float]
    field_id: str
    font: str = "Unknown"
    font_size: float = 0.0


class TransactionCoordinatesSchema(BaseModel):
    date: FieldCoordinateSchema | None = None
    description: FieldCoordinateSchema | None = None
    debit: FieldCoordinateSchema | None = None
    credit: FieldCoordinateSchema | None = None
    balance: FieldCoordinateSchema | None = None


class TransactionResponseSchema(BaseModel):
    transaction_id: str
    page: int
    row_index: int
    date: str | None = None
    description: str = ""
    debit: Decimal | None = None
    credit: Decimal | None = None
    balance: Decimal | None = None
    coordinates: TransactionCoordinatesSchema
    font_metadata: dict[str, Any] = Field(default_factory=dict)
    row_bbox: list[float] = Field(default_factory=list)
    confidence: float = 1.0
    validation_warnings: list[str] = Field(default_factory=list)


class TransactionsListResponse(BaseModel):
    statement_id: UUID
    bank: str
    bank_confidence: float
    transactions: list[TransactionResponseSchema]
    summary: TransactionSummary
    cached: bool = False
    warnings: list[str] = Field(default_factory=list)
    debug: ParseDebugInfo | None = None
    extraction_mode: str = "native"
    layout_confidence: float | None = None
    ocr_confidence: float | None = None
